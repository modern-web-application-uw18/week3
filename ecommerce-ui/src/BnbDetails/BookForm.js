import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './BookForm.css';

class BookForm extends Component {

  constructor(props) {
    super(props);
    this.state = {value: ''};

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange = (event) => {
    this.setState({value: event.target.value});
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const purchase = {
      title: this.props.bnb.title,
      houseType: this.props.bnb.houseType,
      image: this.props.bnb.image,
      city: this.props.bnb.location.city,
      country: this.props.bnb.location.country,
      paymentCost: this.props.bnb.payment.cost,
      paymentDescription: this.props.bnb.payment.description,
      stay: this.state.value
    };
    this.props.addItemToCart(purchase);
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit} className="book-form">
        <label>
          <span>When do you want to stay?</span>
          <input type="date" value={this.state.value} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Add to Cart" />
      </form>
    );
  }
}

export default BookForm;

BookForm.propTypes = {
  addItemToCart: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  houseType: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  location: PropTypes.object.isRequired,
  location: PropTypes.shape({
    city: PropTypes.string,
    country: PropTypes.string
  }),
  payment: PropTypes.object,
  payment: PropTypes.shape({
    cost: PropTypes.number,
    description: PropTypes.string
  })
};
