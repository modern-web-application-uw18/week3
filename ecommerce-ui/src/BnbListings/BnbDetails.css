@import './BnbListings.css';

.bnb-details {
  display: flex;
}

.bnb-details .media {
  max-width: 32rem;
}

.bnb-details .media img {
  width: 100%;
  border: .5px solid rgba(0,0,0,.1);
}

.bnb-details .content {
  padding: 1rem;
}

.bnb-details .bnb-stats {
  margin-bottom: 1.5rem;
}

@media screen and (max-width: 640px) {
  .bnb-details {
    flex-direction: column;
  }
}

@media screen and (min-width: 641px) {
  .bnb-details {
    flex-direction: row;
  }
}
