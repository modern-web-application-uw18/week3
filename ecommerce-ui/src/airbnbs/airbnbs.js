import React, { Component } from 'react';
import airbnbs from './airbnbs.json';
import AddRental from '../addRental/addRental';
import Cart from '../cart/cart';
import '../App.css';
import CartTotal from '../cartTotal/cartTotal';

class Airbnblocation extends Component{
    constructor(props){
        super(props);
        this.state= {
            locations: airbnbs,
            cart:[],
        };
    }


  addRental = (data) => {
      this.setState((prevState, props) => {
          const cart = data;
          const newCart = prevState.cart;
          newCart.push(cart);
          return {
              cart: newCart,
          };
      });
  }

   deleteRental = (index) => {
    this.setState((prevState, props) => {
        const newCart = prevState.cart;
        newCart.splice(index, 1);
        return {
            cart: newCart
        };
    });
}

      render() {
        let locations = this.state.locations.map((locations, idx) => {
          return(
              <AddRental key={idx}
              addRental={this.addRental}
              image={locations.image}
              title={locations.title}
              stars={locations.rating.stars}
              numberofreviews={locations.rating.reviews}
              houseType={locations.houseType}
              city={locations.location.city}
              country={locations.location.country}
              cost={locations.payment.cost}
              description={locations.payment.description}
              host={locations.host.name}
              issuperhost={locations.host.isSuperhost}
              />
            )
      })
      let shoppingCart = this.state.cart.map((cart, idx) => {   
          return(
            <Cart key={idx}
            index={idx}
            deleteRental={this.deleteRental}
            image={cart.image}
            title={cart.title}
            stars={cart.stars}
            numberofreviews={cart.reviews}
            houseType={cart.houseType}
            city={cart.city}
            country={cart.country}
            cost={cart.cost}
            description={cart.description}
            host={cart.name}
            issuperhost={cart.isSuperhost}
            />
          )
    })
      return (

        <div>
          <h3>Cart</h3>
          <CartTotal
          cost={this.state.cart}/>
          <div  className="flex-container">
          {shoppingCart}
          </div>
          <h3>Pick your destination!</h3>
          <div className="flex-container">{locations}</div>
        </div>

      )
    }
    }

    export default Airbnblocation;
