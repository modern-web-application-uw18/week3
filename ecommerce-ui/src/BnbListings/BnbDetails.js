import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './BnbDetails.css';
import airbnbs from '../airbnbs.json';
import SuperHostIcon from './SuperHostIcon';
import HostIcon from './HostIcon';
import BookForm from './BookForm';

class BnbDetails extends Component {
  constructor(props) {
    super(props);
  }

  getHostIcon() {
    return (
      <div className="host-icon">
        <HostIcon />
      </div>
    );
  }

  getSuperHostIcon() {
    return (
      <div className="host-icon">
        <SuperHostIcon />
      </div>
    );
  }

  getPaymentDescription(descript) {
    return (
      <span>
        · {descript}
      </span>
    );
  };

  render() {
    const itemID = this.props.match.params.id;
    const obj = airbnbs.find(function (obj) { return obj.id === itemID;});
    const isSuperHost = obj.host.isSuperhost;
    const paymentDescription = obj.payment.description;
    return (
      <section className="bnb-details">
        <div className="media">
          <img src={obj.image} alt={obj.title} title={obj.title}/>
        </div>
        <div className="content">
        <h2>{obj.houseType} · {obj.location.city}, {obj.location.country}</h2>
        <h1>{obj.title}</h1>
        <p><span className="price">${obj.payment.cost}</span> per night  {paymentDescription ? this.getPaymentDescription(paymentDescription) : null}</p>
        <div className="bnb-stats">
          <div className="rating-data">
            <p>Rating: {obj.rating.stars} stars / {obj.rating.reviews} reviews</p>
          </div>
          <div className="host-info" alt={obj.host.name} title={obj.host.name}>
            {isSuperHost ? this.getSuperHostIcon() : this.getHostIcon() }
            <a href="/">{obj.host.name}</a>
          </div>
        </div>
        <BookForm bnb={obj} addItemToCart={this.props.addItemToCart}/>
        </div>
      </section>

    );
  }
}

export default BnbDetails;

BnbDetails.propTypes = {
  addItemToCart: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  houseType: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  location: PropTypes.object.isRequired,
  location: PropTypes.shape({
    city: PropTypes.string,
    country: PropTypes.string
  }),
  payment: PropTypes.object,
  payment: PropTypes.shape({
    cost: PropTypes.number,
    description: PropTypes.string
  }),
  host: PropTypes.object,
  host: PropTypes.shape({
    name: PropTypes.string,
    isSuperhost: PropTypes.bool
  }),
  rating: PropTypes.object,
  rating: PropTypes.shape({
    stars: PropTypes.number,
    reviews: PropTypes.number
  })
};
