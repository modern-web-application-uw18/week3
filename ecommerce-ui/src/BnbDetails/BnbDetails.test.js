import React from 'react';
import BnbDetails from './BnbDetails';
import ReactTestRenderer from 'react-test-renderer';
import airbnbs from '../airbnbs.json';
import PropTypes from 'prop-types'


export default class WithContext extends React.Component {
  static contextTypes = {
    match: PropTypes.any,
    addItemToCart: PropTypes.any,
  }

  render () {
    const match = {
      params: {
        id: airbnbs[0].id
      }
    }
    return (
      <div><BnbDetails match={match} addItemToCart={"test"} /></div>
    )
  }
}

describe('BnbDetails component', () => {
  it('should render', () => {
    const component = ReactTestRenderer.create(<WithContext />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
