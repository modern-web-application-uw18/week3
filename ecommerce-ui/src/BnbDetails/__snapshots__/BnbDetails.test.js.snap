// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BnbDetails component should render 1`] = `
<div>
  <section
    className="bnb-details"
  >
    <div
      className="media"
    >
      <img
        alt="Centrally-located Manhattan studio"
        src="https://images.unsplash.com/photo-1490806230066-f7e6f7bf5052?ixlib=rb-0.3.5&ixid=eyJhcHBfaWQiOjEyMDd9&s=2ca889535ea01f912f94ac4ddf0034e0&auto=format&fit=crop&w=500&q=80"
        title="Centrally-located Manhattan studio"
      />
    </div>
    <div
      className="content"
    >
      <h2>
        Apartment
         · 
        New York
        , 
        USA
      </h2>
      <h1>
        Centrally-located Manhattan studio
      </h1>
      <p>
        <span
          className="price"
        >
          $
          105
        </span>
         per night  
        <span>
          · 
          Free cancellation
        </span>
      </p>
      <div
        className="bnb-stats"
      >
        <div
          className="rating-data"
        >
          <p>
            Rating: 
            3
             stars / 
            133
             reviews
          </p>
        </div>
        <div
          alt="Robert"
          className="host-info"
          title="Robert"
        >
          <div
            className="host-icon"
          >
            <svg
              fill="none"
              height="24"
              stroke="rgb(130, 110, 85)"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="8.5"
                cy="7"
                r="4"
              />
              <line
                x1="20"
                x2="20"
                y1="8"
                y2="14"
              />
              <line
                x1="23"
                x2="17"
                y1="11"
                y2="11"
              />
            </svg>
          </div>
          <a
            href="/"
          >
            Robert
          </a>
        </div>
      </div>
      <form
        className="book-form"
        onSubmit={[Function]}
      >
        <label>
          <span>
            When do you want to stay?
          </span>
          <input
            onChange={[Function]}
            type="date"
            value=""
          />
        </label>
        <input
          type="submit"
          value="Add to Cart"
        />
      </form>
    </div>
  </section>
</div>
`;
